File: apps/api/app/schemas/attack.py
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field


class AttackResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    episode_id: UUID
    technique_id: str
    owasp_category: str
    attack_type: str
    success: bool
    evidence: dict
    confidence: float
    payload: Optional[str] = None
    target_endpoint: Optional[str] = None
    http_method: Optional[str] = None
    request_headers: Optional[dict] = None
    request_body: Optional[str] = None
    response_status: Optional[int] = None
    response_body: Optional[str] = None
    timestamp: datetime
    created_at: datetime


class AttackListParams(BaseModel):
    episode_id: UUID
    technique_id: Optional[str] = None
    owasp_category: Optional[str] = None
    success: Optional[bool] = None
    date_from: Optional[datetime] = None
    date_to: Optional[datetime] = None
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=50, ge=1, le=100)


class AttackStats(BaseModel):
    total_attacks: int = 0
    successful_attacks: int = 0
    failed_attacks: int = 0
    by_technique: dict = Field(default_factory=dict)
    by_owasp_category: dict = Field(default_factory=dict)

File: packages/shared/src/schemas/attack.ts
import { z } from "zod";
import { UUIDSchema, DateTimeSchema } from "./common";

export const AttackSchema = z.object({
  id: UUIDSchema,
  episodeId: UUIDSchema,
  techniqueId: z.string(),
  owaspCategory: z.string(),
  attackType: z.string(),
  success: z.boolean(),
  evidence: z.record(z.unknown()).default({}),
  confidence: z.number().min(0).max(1),
  payload: z.string().optional(),
  targetEndpoint: z.string().optional(),
  httpMethod: z.string().optional(),
  requestHeaders: z.record(z.string()).optional(),
  requestBody: z.string().optional(),
  responseStatus: z.number().int().optional(),
  responseBody: z.string().optional(),
  timestamp: DateTimeSchema,
  createdAt: DateTimeSchema
});
export type Attack = z.infer<typeof AttackSchema>;

export const AttackListParamsSchema = z.object({
  episodeId: UUIDSchema,
  techniqueId: z.string().optional(),
  owaspCategory: z.string().optional(),
  success: z.boolean().optional(),
  dateFrom: DateTimeSchema.optional(),
  dateTo: DateTimeSchema.optional(),
  page: z.number().int().positive().default(1),
  limit: z.number().int().positive().max(100).default(50)
});
export type AttackListParams = z.infer<typeof AttackListParamsSchema>;

export const AttackStatsSchema = z.object({
  totalAttacks: z.number().int().nonnegative(),
  successfulAttacks: z.number().int().nonnegative(),
  failedAttacks: z.number().int().nonnegative(),
  byTechnique: z.record(z.object({
    total: z.number().int().nonnegative(),
    successful: z.number().int().nonnegative()
  })),
  byOwaspCategory: z.record(z.object({
    total: z.number().int().nonnegative(),
    successful: z.number().int().nonnegative()
  }))
});
export type AttackStats = z.infer<typeof AttackStatsSchema>;